File: cwt-spring/src/main/java/com/cwtsite/cwt/domain/game/entity/Game.java
package com.cwtsite.cwt.domain.game.entity;

import com.cwtsite.cwt.entity.Comment;
import com.cwtsite.cwt.domain.group.entity.Group;
import com.cwtsite.cwt.domain.tournament.entity.Tournament;
import com.cwtsite.cwt.domain.user.repository.entity.User;
import com.fasterxml.jackson.annotation.JsonIgnore;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import javax.persistence.*;
import java.io.Serializable;
import java.sql.Timestamp;
import java.util.List;
import java.util.Objects;

/**
 * @todo There should definitely be more lazy loading here.
 */
@Entity
@Table(name = "game")
@SequenceGenerator(name = "game_seq", sequenceName = "game_seq", initialValue = 1219, allocationSize = 1)
public class Game implements Serializable {

    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "game_seq")
    private Long id;

    @Column(name = "score_home")
    private Integer scoreHome;

    @Column(name = "score_away")
    private Integer scoreAway;

    @Column(name = "tech_win")
    private Boolean techWin = false;

    @Column(name = "downloads")
    private Integer downloads;

    @Column(name = "created")
    @CreationTimestamp
    private Timestamp created;

    @Column(name = "modified")
    @UpdateTimestamp
    private Timestamp modified;

    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(unique = true)
    private PlayoffGame playoff;

    @ManyToOne
    private Tournament tournament;

    @JsonIgnore
    @ManyToOne
    @JoinColumn(name = "group_id")
    private Group group;

    @ManyToOne
    private User homeUser;

    @ManyToOne
    private User awayUser;

    @ManyToOne
    private User reporter;

    @OneToMany(cascade = CascadeType.ALL, mappedBy = "game")
    private List<Rating> ratings;

    @OneToMany(cascade = CascadeType.ALL, mappedBy = "game")
    private List<Comment> comments;

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Integer getScoreHome() {
        return scoreHome;
    }

    public void setScoreHome(Integer scoreHome) {
        this.scoreHome = scoreHome;
    }

    public Game scoreHome(Integer scoreHome) {
        this.scoreHome = scoreHome;
        return this;
    }

    public Integer getScoreAway() {
        return scoreAway;
    }

    public void setScoreAway(Integer scoreAway) {
        this.scoreAway = scoreAway;
    }

    public Game scoreAway(Integer scoreAway) {
        this.scoreAway = scoreAway;
        return this;
    }

    public Boolean isTechWin() {
        return techWin;
    }

    public Game techWin(Boolean techWin) {
        this.techWin = techWin;
        return this;
    }

    public void setTechWin(Boolean techWin) {
        this.techWin = techWin;
    }

    public Integer getDownloads() {
        return downloads;
    }

    public void setDownloads(Integer downloads) {
        this.downloads = downloads;
    }

    public Game downloads(Integer downloads) {
        this.downloads = downloads;
        return this;
    }

    public Timestamp getCreated() {
        return created;
    }

    public void setCreated(Timestamp created) {
        this.created = created;
    }

    public Game created(Timestamp created) {
        this.created = created;
        return this;
    }

    public Timestamp getModified() {
        return modified;
    }

    public void setModified(Timestamp modified) {
        this.modified = modified;
    }

    public Game modified(Timestamp modified) {
        this.modified = modified;
        return this;
    }

    public PlayoffGame getPlayoff() {
        return playoff;
    }

    public void setPlayoff(PlayoffGame playoffGame) {
        this.playoff = playoffGame;
    }

    public Game playoff(PlayoffGame playoffGame) {
        this.playoff = playoffGame;
        return this;
    }

    public Tournament getTournament() {
        return tournament;
    }

    public void setTournament(Tournament tournament) {
        this.tournament = tournament;
    }

    public Game tournament(Tournament tournament) {
        this.tournament = tournament;
        return this;
    }

    public Group getGroup() {
        return group;
    }

    public void setGroup(Group group) {
        this.group = group;
    }

    public Game group(Group group) {
        this.group = group;
        return this;
    }

    public User getHomeUser() {
        return homeUser;
    }

    public void setHomeUser(User user) {
        this.homeUser = user;
    }

    public Game homeUser(User user) {
        this.homeUser = user;
        return this;
    }

    public User getAwayUser() {
        return awayUser;
    }

    public void setAwayUser(User user) {
        this.awayUser = user;
    }

    public Game awayUser(User user) {
        this.awayUser = user;
        return this;
    }

    public User getReporter() {
        return reporter;
    }

    public void setReporter(User user) {
        this.reporter = user;
    }

    public Game reporter(User user) {
        this.reporter = user;
        return this;
    }

    public List<Rating> getRatings() {
        return ratings;
    }

    public void setRatings(List<Rating> ratings) {
        this.ratings = ratings;
    }

    public List<Comment> getComments() {
        return comments;
    }

    public void setComments(List<Comment> comments) {
        this.comments = comments;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Game game = (Game) o;
        if (game.id == null || id == null) {
            return false;
        }
        return Objects.equals(id, game.id);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id);
    }

    @Override
    public String toString() {
        return "Game{" +
                "id=" + id +
                ", scoreHome='" + scoreHome + "'" +
                ", scoreAway='" + scoreAway + "'" +
                ", techWin='" + techWin + "'" +
                ", downloads='" + downloads + "'" +
                ", created='" + created + "'" +
                ", modified='" + modified + "'" +
                '}';
    }
}

File: cwt-spring/src/main/java/com/cwtsite/cwt/domain/game/entity/Game.kt
package com.cwtsite.cwt.domain.game.entity

import com.cwtsite.cwt.domain.bet.entity.Bet
import com.cwtsite.cwt.domain.group.entity.Group
import com.cwtsite.cwt.domain.tournament.entity.Tournament
import com.cwtsite.cwt.domain.user.repository.entity.User
import com.cwtsite.cwt.entity.Comment
import com.fasterxml.jackson.annotation.JsonIgnore
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.Formula
import org.hibernate.annotations.UpdateTimestamp
import java.sql.Timestamp
import javax.persistence.*

@Entity
@Table(name = "game")
@SequenceGenerator(name = "game_seq", sequenceName = "game_id_seq", allocationSize = 5)
data class Game(

        @Id
        @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "game_seq")
        var id: Long? = null,

        @Column(name = "score_home")
        var scoreHome: Int? = null,

        @Column(name = "score_away")
        var scoreAway: Int? = null,

        @Column(name = "tech_win")
        var techWin: Boolean = false,

        @Column(name = "reported_at")
        var reportedAt: Timestamp? = null,

        @Column(name = "replay_quantity")
        var replayQuantity: Int? = null,

        @field:CreationTimestamp
        @Column(name = "created", nullable = false, updatable = false)
        var created: Timestamp? = null,

        @field:UpdateTimestamp
        @Column(name = "modified", nullable = false)
        var modified: Timestamp? = null,

        @OneToOne(cascade = [CascadeType.ALL])
        @JoinColumn(unique = true)
        var playoff: PlayoffGame? = null,

        @ManyToOne(optional = false)
        @JoinColumn(nullable = false)
        var tournament: Tournament,

        @JsonIgnore
        @ManyToOne
        @JoinColumn(name = "group_id")
        var group: Group? = null,

        @ManyToOne
        var homeUser: User? = null,

        @ManyToOne
        var awayUser: User? = null,

        @ManyToOne
        @JoinColumn
        var reporter: User? = null,

        @OneToMany(cascade = [CascadeType.ALL], mappedBy = "game")
        val ratings: List<Rating> = mutableListOf(),

        @OneToMany(cascade = [CascadeType.ALL], mappedBy = "game")
        val comments: List<Comment> = mutableListOf(),

        @OneToMany(cascade = [CascadeType.ALL], mappedBy = "game")
        val bets: List<Bet> = mutableListOf(),

        @JsonIgnore
        @OneToOne(fetch = FetchType.LAZY, cascade = [CascadeType.ALL], orphanRemoval = true)
        @JoinColumn(unique = true)
        var replay: Replay? = null,

        @Basic(fetch = FetchType.LAZY)
        @Formula("(select count(*) from COMMENT c where c.GAME_ID = id)")
        val commentsSize: Int? = null,

        @Basic(fetch = FetchType.LAZY)
        @Formula("(select count(*) from RATING r where r.GAME_ID = id)")
        val ratingsSize: Int? = null,

        var voided: Boolean = false
) {

    fun pairingInvolves(user: User?) = (user != null) && (homeUser == user || awayUser == user)

    fun pairingCompleted() = homeUser != null && awayUser != null

    fun pairingHalfCompleted() = (homeUser == null) xor (awayUser == null)

    fun pairingEmpty() = homeUser == null && awayUser == null

    fun pairUser(user: User) {
        if (!pairingHalfCompleted()) throw RuntimeException("Game is not half paired.")
        if (homeUser == null) homeUser = user else awayUser = user
    }

    fun wasPlayed() = scoreHome != null && scoreAway != null && homeUser != null && awayUser != null

    fun winner() = (if (scoreHome!! > scoreAway!!) homeUser else awayUser)!!

    fun loser() = (if (scoreHome!! > scoreAway!!) awayUser else homeUser)!!

    fun playoff() = playoff != null && group == null

    fun group() = !playoff()

    override fun toString(): String {
        return "Game(id=$id)"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Game

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id?.hashCode() ?: 0
    }
}
